build/sdk_update.go

```diff
@@ -0,0 +1,316 @@
+
+	t, err := template.New(manifestTpl).Funcs(funcs).Parse(tpl)
+	if err != nil {
+		return err
+	}
+
+	name := fixGitFolder(manifestTpl)
+	file := filepath.Join(root, manifest.Filename)
+
+	if _, err = os.Stat(file); err == nil {
+		return errors.New("trying to overwrite the manifest")
+	}
+	info, _ := skeleton.AssetInfo(name)
+
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_TRUNC, info.Mode())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	args := map[string]string{
+		"Language": language,
+	}
+	if err := t.Execute(f, args); err != nil {
+		return err
+	}
+	return f.Close()
+}
+
+// SDKUpdate updates SDK-managed files for the driver located at root.
+//
+// If DryRun option is set, the function would not update any files, and instead will
+// return ErrChangesRequired if there are any changes required.
+func SDKUpdate(root string, opt *UpdateOptions) error {
+	if opt == nil {
+		opt = &UpdateOptions{}
+	}
+	if opt.Warningf == nil {
+		opt.Warningf = Silencef
+	}
+	if opt.Noticef == nil {
+		opt.Noticef = Silencef
+	}
+	if opt.Debugf == nil {
+		opt.Debugf = Silencef
+	}
+
+	m, err := manifest.Load(filepath.Join(root, manifest.Filename))
+	if err != nil {
+		return err
+	}
+	c := &updater{
+		root: root, opt: *opt,
+		context: map[string]interface{}{
+			"Manifest": m,
+		},
+	}
+
+	for _, file := range skeleton.AssetNames() {
+		if file == manifestTpl {
+			continue
+		}
+
+		if err := c.processAsset(file); err != nil {
+			return err
+		}
+	}
+	d, err := NewDriver(root)
+	if err != nil {
+		return err
+	}
+	var changed bool
+	if c.opt.DryRun {
+		changed, err = d.ScriptChanged()
+	} else {
+		changed, err = d.Prepare()
+	}
+	if err != nil {
+		return err
+	} else if changed {
+		c.notifyChangedFile(ScriptName)
+	}
+
+	if c.opt.DryRun && c.changes > 0 {
+		return ErrChangesRequired
+	}
+
+	return nil
+}
+
+func (c *updater) processAsset(name string) error {
+	overwrite := managedFiles[name]
+	if overwriteManagedFiles {
+		overwrite = false
+	}
+
+	if strings.HasSuffix(name, tplExt) {
+		return c.processTemplateAsset(name, c.context, overwrite)
+	}
+
+	return c.processFileAsset(name, overwrite)
+}
+
+func (c *updater) processFileAsset(name string, overwrite bool) error {
+	content := skeleton.MustAsset(name)
+	info, _ := skeleton.AssetInfo(name)
+
+	name = fixGitFolder(name)
+	return c.writeFile(filepath.Join(c.root, name), content, info.Mode(), overwrite)
+}
+
+var funcs = map[string]interface{}{
+	"escape_shield": escapeShield,
+	"expName":       expName,
+}
+
+func (c *updater) processTemplateAsset(name string, v interface{}, overwrite bool) error {
+	tpl := string(skeleton.MustAsset(name))
+
+	t, err := template.New(name).Funcs(funcs).Parse(tpl)
+	if err != nil {
+		return err
+	}
+
+	name = fixGitFolder(name)
+	file := filepath.Join(c.root, name[:len(name)-len(tplExt)])
+
+	buf := bytes.NewBuffer(nil)
+	if err := t.Execute(buf, v); err != nil {
+		return err
+	}
+
+	info, _ := skeleton.AssetInfo(name)
```

Comment:
We should probably not ignore these errors, but it'd be fine to panic for them if indeed this is the mirror of `skeleton.MustAsset` for the file info instead of the content:

```go
func mustAssetInfo(name string) os.FileInfo {
  fi, err := skeleton.AssetInfo(name)
  if err != nil {
    log.Panicf("Missing asset info for %q: %v", name, err)
  }
  return fi
}
```

or whatever.

(In fact we might think about adding that to the API for the `skeleton` package separately, but for now I think it'd be fine to host it here)